### PR TITLE
feat: hermetic build OS detection

### PR DESCRIPTION
**Pull request by diegomarquezp**
Adds OS detection to `generate_library.sh`

Manual input from workflow matrix is removed in favor of automatic detection

### CODE REVIEW

**Comment by diegomarquezp** [comment]
[Mismatch in `diff`](https://github.com/googleapis/sdk-platform-java/actions/runs/6175923393/job/16763775744?pr=1988) due to `googleapis` recently pushing a [commit](https://github.com/googleapis/googleapis/commit/e8a136feaca2547dd5566ef79841d28f76a80eb5) not found in `googleapis-gen` ([link](https://github.com/googleapis/googleapis-gen/tree/b19bf82fcb471fe7a53d107528dbf5b78eee26e4/google/bigtable/v2))

File: library_generation/utilities.sh [hunk 1]
```diff
@@ -273,3 +273,19 @@ get_version_from_versions_txt() {
   version=$(grep "$key:" "${versions}" | cut -d: -f3) # 3rd field is snapshot
   echo "${version}"
 }
+
+detect_OS() {
```
Comment (JoeWang1127) [inline library_generation/utilities.sh:277]: Judging by what the function does, the name should be `detect_os_architecture`

File: library_generation/utilities.sh [hunk 2]
```diff
@@ -273,3 +273,19 @@ get_version_from_versions_txt() {
...
+  if [[ "${OSTYPE}" == "linux-gnu"* ]] || [[ "${OSTYPE}" == "freebsd"* ]]; then
+    os_architecture="linux-x86_64"
+  elif [[ "${OSTYPE}" == "darwin"* ]]; then
```
Comment (JoeWang1127) [inline library_generation/utilities.sh:280]: What about `osx-aarm_64`?

I think you can use `uname -sm` to detect OS type and CPU architecture, but I'm not sure whether windows can run this command.

File: library_generation/utilities.sh [hunk 3]
```diff
@@ -273,3 +273,19 @@ get_version_from_versions_txt() {
   version=$(grep "$key:" "${versions}" | cut -d: -f3) # 3rd field is snapshot
   echo "${version}"
 }
+
+detect_OS() {
+  if [[ "${OSTYPE}" == "linux-gnu"* ]] || [[ "${OSTYPE}" == "freebsd"* ]]; then
```
Comment (JoeWang1127) [inline library_generation/utilities.sh:278]: Maybe replace `if` with `case`?

**Review by JoeWang1127 — CHANGES_REQUESTED** [review]

File: library_generation/utilities.sh (hunk 1 above)
Comment (diegomarquezp) [reply library_generation/utilities.sh]: makes more sense. Changed

File: library_generation/utilities.sh (hunk 3 above)
Comment (diegomarquezp) [reply library_generation/utilities.sh]: looks better, changed

File: library_generation/utilities.sh (hunk 2 above)
Comment (diegomarquezp) [reply library_generation/utilities.sh]: I defaulted to `win32` regardless of architecture and changed it to `uname -m` to detect the architecture only, while keep reliance on `${OSTYPE}`

File: library_generation/utilities.sh [hunk 4]
```diff
@@ -273,3 +273,24 @@ get_version_from_versions_txt() {
   version=$(grep "$key:" "${versions}" | cut -d: -f3) # 3rd field is snapshot
   echo "${version}"
 }
+
+detect_os_architecture() {
+  case "${OSTYPE}" in
```
Comment (JoeWang1127) [inline library_generation/utilities.sh:278]: where does `OSTYPE` come from?

File: library_generation/utilities.sh [hunk 5]
```diff
@@ -273,3 +273,24 @@ get_version_from_versions_txt() {
...
+    "linux-gnu"*|"freebsd"*)
+      os_architecture="linux-$(uname -m)"
+      ;;
+    "darwin"*)
+      os_architecture="osx-$(uname -m)"
```
Comment (JoeWang1127) [inline library_generation/utilities.sh:283]: When I run `uname -m` on my machine, the output is `arm64`.
However, the right os architecture should be osx-aarch_64.
You can reference the os architecture [here](https://github.com/protocolbuffers/protobuf/releases)

I also implemented a os architecture detection function:
```
get_os_architecture() {
  local os_type
  local os_architecture
  os_type=$(uname -sm)
  case "${os_type}" in
    *"Linux x86_64"*)
      os_architecture="linux-x86_64"
      ;;
    *"Darwin x86_64"*)
      os_architecture="osx-x86_64"
      ;;
    *)
      os_architecture="osx-aarch_64"
      ;;
  esac
  echo "${os_architecture}"
}
```

Comment (diegomarquezp) [reply library_generation/utilities.sh]: ah, thanks for the function :) I couldn't confirm for the `aarch64` case, but I switched the detection function to this one

**Review by JoeWang1127 — APPROVED** [review]

**Pull request merged**